security/selinux.ksy

```diff
@@ -0,0 +1,1605 @@
+      type: xen_ioport
+      repeat: expr
+      repeat-expr: elements_count
+      if: elements_count > 0
+
+  xen_ioport:
+    seq:
+    - id: low_port
+      type: u4
+    - id: high_port
+      type: u4
+    - id: context0
+      type: context
+
+  xen_iomems:
+    seq:
+    - id: elements_count
+      -orig-id: nel
+      type: u4
+    - id: xen_iomem
+      type: xen_iomem
+      repeat: expr
+      repeat-expr: elements_count
+      if: elements_count > 0
+
+  xen_iomem:
+    seq:
+    - id: low_iomem64
+      type: u8
+      if: '_root.version >= 30'
+    - id: high_iomem64
+      type: u8
+      if: '_root.version >= 30'
+    - id: low_iomem32
+      type: u4
+      if: '_root.version < 30'
+    - id: high_iomem32
+      type: u4
+      if: '_root.version < 30'
+    - id: context0
+      type: context
+
+  xen_pcidevices:
+    seq:
+    - id: elements_count
+      -orig-id: nel
+      type: u4
+    - id: xen_pcidevice
+      type: xen_pcidevice
+      repeat: expr
+      repeat-expr: elements_count
+      if: elements_count > 0
+
+  xen_pcidevice:
+    seq:
+    - id: device
+      type: u4
+    - id: context0
+      type: context
+
+
+  xen_devicetrees:
+    seq:
+    - id: elements_count
+      -orig-id: nel
+      type: u4
+    - id: xen_devicetree
+      type: xen_devicetree
+      repeat: expr
+      repeat-expr: elements_count
+      if: elements_count > 0
+
+  xen_devicetree:
+    seq:
+    - id: length
+      -orig-id: len
+      type: u4
+      valid:
+        min: 1
+    - id: name
+      type: str
+      encoding: ascii
+      size: length
+    - id: context0
+      type: context
+
+  genfs: # genfs_read
+    seq:
+    - id: elements_count
+      -orig-id: nel
+      type: u4
+    - id: items
+      type: genfs_item
+      repeat: expr
+      repeat-expr: elements_count
+      if: elements_count > 0
+
+  genfs_item:
+    seq:
+    - id: length
+      -orig-id: len
+      type: u4
+      valid:
+        min: 1
+    - id: fstype
+      type: str
+      encoding: ascii
+      size: length
+    - id: elements_count
+      -orig-id: nel2
+      type: u4
+    - id: items
+      type: genfs2_item
+      repeat: expr
+      repeat-expr: elements_count
+      if: elements_count > 0
+
+  genfs2_item:
+    seq:
+    - id: length
+      -orig-id: len
+      type: u4
+      valid:
+        min: 1
+    - id: name
+      type: str
+      encoding: ascii
+      size: length
+    - id: sclass
+      type: u4
+    - id: context0
+      type: context
+
+  range: # range_read
+    seq:
+    - id: elements_count
+      -orig-id: nel
+      type: u4
+    - id: items
+      type: range_item
+      repeat: expr
+      repeat-expr: elements_count
+      if: elements_count > 0
+
+  range_item:
+    seq:
+    - id: source_type
+      type: u4
+    - id: target_type
+      type: u4
+    - id: target_class
+      type: u4
+      if: '_root.type == policy_types::kernel and _root.version >= 21'
+    - id: range_tr
+      type: mls_range
+
+
+instances:
```

Comment:
Hmm, although the [Style Guide](https://doc.kaitai.io/ksy_style_guide.html#type) isn't too specific in this (I thought it was), I'd definitely put `instances` before `types` in this case, i.e. here:

https://github.com/kaitai-io/kaitai_struct_formats/blob/3b71b9fe4640eb00a48c255baf4d6cfdbbe73828/security/selinux.ksy#L11-L15

```diff
 seq:
   - id: header
     type: header
+
+instances:
+  ...

 types:
```

This helps with readability a lot in my opinion. First, you fully define the attributes of the root type (i.e. `seq` and `instances`, which are not all that different - they're accessed in exactly the same way) and then you provide definitions of other types.